### PR TITLE
Start reimplementation of simplex pattern

### PR DESCRIPTION
**Pull request by SirYwell**
## Overview
Would partly cover #437 (the simplex mask isn't included)

### Short version: 
I've added a SimplexPatternParser which parses `#simplex[scale][pattern]`.
### Long version: 
As FAWE has multiple patterns/masks that require additional arguments (provided in [] normally), I've added the class `RichParser` which allows to easily parse those arguments.  Other parsers can extend it, they only need to provide the name of the pattern or mask. They then need to implement the following methods:
##### - `protected Stream<String> getSuggestions(String argument, int index)` 
(maybe that should be optional) For suggestions of the argument at the given index (e.g. for the simplex pattern, it will suggest numbers as first (index 0) argument and patterns as second (index 1) argument.

##### - `protected Pattern parseFromInput(String[] arguments, ParserContext context)`
Parses the actual pattern or mask with the given arguments. The arguments can be validated there.

That way, the parsers don't need to handle the extraction of arguments as this is done by the RichParser (If you have a better name for it, please let me know).

~This is a draft as it's not ready to merge. While the simplex pattern already works, there are still code parts that need to be discussed. I'll comment those issues.~

## Checklist
- [x] I included all information required in the sections above
- [x] I tested my changes and approved their functionality
- [ ] I ensured my changes do not break other parts of the code
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/FastAsyncWorldEdit/blob/1.15/CONTRIBUTING.md)

### CODE REVIEW

File: worldedit-core/src/main/java/com/sk89q/util/StringUtil.java [hunk 1]
```diff
@@ -331,4 +332,37 @@ public static int getLevenshteinDistance(String s, String t) {
...
+    /**
+     * Splits a string respecting enclosing quotes.
+     *
+     * @param input the input to split.
+     * @param delimiter the delimiter to split on.
+     * @param open the opening quote character.
+     * @param close the closing quote character.
+     * @return a list of split strings.
+     */
+    public static List<String> split(String input, char delimiter, char open, char close) {
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/util/StringUtil.java:345]: This one is most likely to break other things. WE uses the method above this one to parse strings but it doesn't work with nested [] brackets. That broke using simplex in simplex

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/pattern/RandomPattern.java [hunk 2]
```diff
@@ -52,6 +52,19 @@ public RandomPattern(SimpleRandom random) {
         this.random = random;
     }
 
+    /**
+     * Create a random pattern from an existing one but with a different random.
+     *
+     * @param random the new random to use.
+     * @param parent the existing random pattern.
+     */
+    public RandomPattern(SimpleRandom random, RandomPattern parent) {
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/pattern/RandomPattern.java:61]: We either need this one or make the random non-final, reassign it and recreate the RandomCollection with it

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            throw new InputParseException("Invalid amount of arguments. Syntax: #simplex[scale][pattern]");
+        }
+        double scale = Double.parseDouble(arguments[0]);
+        scale = 1d / Math.max(1, scale);
+        Pattern inner = worldEdit.getPatternFactory().parseFromInput(arguments[1], context);
+        if (inner instanceof RandomPattern) {
+            return new RandomPattern(new SimplexRandom(scale), (RandomPattern) inner);
+        }
+        // TODO handle RandomStatePattern?
+        // TODO for other patterns, either warn actor (inner pattern will be returned) or throw exception
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java:56]: There might be patterns a simplex can't handle correctly, so we can either just return that pattern (means nothing happens) or throw an exception so the user knows their input was invalid

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if (index == 1) {
+            return worldEdit.getPatternFactory().getSuggestions(argument).stream();
+        }
+        return Stream.empty();
+    }
+
+    @Override
+    protected Pattern parseFromInput(String[] arguments, ParserContext context) {
+        if (arguments.length != 2) {
+            // TODO
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java:46]: I'm not familiar with the way WE handles stuff like syntax error messages, but I guess there's a better way than that

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/RichParser.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        if (StringUtil.split(input, ',', '[', ']').size() > 1) return null;
+        // TODO if no arguments -> exception
+        // TODO if not starts with [ and ends with ] -> exception
+        String[] arguments = extractArguments(input.substring(prefix.length()), true);
+        return parseFromInput(arguments, context);
+    }
+
+    protected abstract Stream<String> getSuggestions(String argument, int index);
+
+    protected abstract E parseFromInput(String[] arguments, ParserContext context);
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/RichParser.java:58]: `@NotNull` annotation is needed for the parameter `arguments`. (Use the JetBrains one)

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java (hunk 4 above)
Comment (MattBDev) [reply worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java:46]: If the length of the input doesn't match the length of the format required then an `InputParseException` should be thrown.
An example would be the following code:
`throw new InputParseException("Clipboard offset needs x,y,z coordinates.");`

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java (hunk 3 above)
Comment (MattBDev) [reply worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java:56]: Use your best judgement here and we can adjust based on feedback.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/pattern/RandomPattern.java (hunk 2 above)
Comment (MattBDev) [reply worldedit-core/src/main/java/com/sk89q/worldedit/function/pattern/RandomPattern.java:61]: Do whatever is easier here.

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java (hunk 3 above)
Comment (SirYwell) [reply worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/pattern/SimplexPatternParser.java]: It's returning inner for `BlockStateHolder`s now as that's single blocks anyways if I got it right. For other patterns than that or RandomPattern, it throws an exception. I played around a lot but if I missed something (that actually makes sense to use in simplex) people can come up with that.

**Pull request merged**